File: Leetcode/Javascript/17. Letter Combinations of a Phone Number.js
var letterCombinations = function(digits) {
    var result = [];
    if(!digits.length) return result;
    var dic = ['',' ','abc','def','ghi','jkl','mno','pqrs','tuv','wxyz'];
    dfs(0, digits, [], dic, result);
    return result;
};
function dfs(i, digits, list, dic, result) {
    if(i === digits.length) {
        result.push(list.join(''));
        return;
    }
    var num = Number.parseInt(digits.charAt(i));
    var candidate = dic[num];
    if(!candidate) dfs(i+1, digits, list, dic, result);
    for(var j = 0; j < candidate.length; ++j) {
        list.push(candidate.charAt(j));
        dfs(i+1, digits, list, dic, result);
        list.pop();
    }
}

File: Leetcode/Three/17. Letter Combinations of a Phone Number.java
public class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> dic = new ArrayList<>();
        dic.add("");
        dic.add(" ");
        dic.add("abc");
        dic.add("def");
        dic.add("ghi");
        dic.add("jkl");
        dic.add("mno");
        dic.add("pqrs");
        dic.add("tuv");
        dic.add("wxyz");
        List<String> result = new ArrayList<>();
        if(digits.length() == 0) return result;
        dfs(result, 0, digits, "", dic);
        return result;
    }
    public void dfs(List<String> result, int index, String digits, String str, List<String> dic) {
        if(index == digits.length()) {
            result.add(str);
            return;
        }
        int x = digits.charAt(index)-'0';
        String s = dic.get(x);
        if(s.length() == 0) dfs(result, index+1, digits, str, dic);
        else {
            for(int i = 0; i < s.length(); ++i) {
                dfs(result, index+1, digits, str+s.charAt(i), dic);
            }
        }
    }
}
